keps/sig-storage/1710-selinux-relabeling/README.md

```diff
@@ -0,0 +1,687 @@
+
+See [SELinux documentation](https://selinuxproject.org/page/NB_MLS) for more details.
+
+### SELinux context assignment
+In Kubernetes, the SELinux context of a pod is assigned in two ways:
+1. Either it is set by user in PodSpec or Container: https://kubernetes.io/docs/tasks/configure-pod-container/security-context/.
+1. If not set in Pod/Container, the container runtime will allocate a new unique SELinux context and assign it to a pod (container) by itself.
+
+### Volumes
+Currently, Kubernetes *knows*(*) which volume plugins supports SELinux (i.e. supports extended attributes on a filesystem the plugin provides).
+If SELinux is supported for a volume, it passes the volume to the container runtime with ":Z" option ("private unshared").
+The container runtime then **recursively relabels** all files on the volume to either the label set in PodSpec/Container or the random value allocated by the container runtime itself.
+
+**This relabeling needs to traverse through the whole volume, and it can be slow for volumes with a large amount of files.**
+
+*) These in-tree volume plugins don't support SELinux: Azure File, CephFS, GlusterFS, HostPath, NFS, Portworx and Quobyte.
+All other volume plugins support it.
+This knowledge is hardcoded in in-tree volume plugins (e.g. [NFS](https://github.com/kubernetes/kubernetes/blob/0c5c3d8bb97d18a2a25977e92b3f7a49074c2ecb/pkg/volume/nfs/nfs.go#L235)).
+
+For CSI, kubelet uses following heuristics:
+
+1. Mount the volume (via `NodeStage` + `NodePublish` CSI calls).
+2. Check mount options of the volume mount dir. If and only if it contains `seclabel` mount option, the volume supports SELinux.
+
+### Goals
+
+Optionally (chosen by user), do not recursively relabel content of the volumes.
+
+### Non-Goals
+
+Change container runtimes / CRI.
+
+## Proposal
+
+Offer option in `Pod.Spec.PodSecurityContext to` *mount* volumes with the right labels instead of recursive relabeling:
+
+```go
+type SELinuxRelabelPolicy string
+
+const (
+    OnVolumeMount SELinuxRelabelPolicy = "OnVolumeMount"
+    AlwaysRelabel SELinuxRelabelPolicy = "Always"
+)
+
+type PodSecurityContext struct {
+    // seLinuxRelabelPolicy ← new field
+    // Defines behavior of changing SELinux labels of the volume before being exposed inside Pod.
+    // Valid values are "OnVolumeMount" and "Always". If not specified, "Always" is used.
+    // "Always" policy recursively changes SELinux labels on all files on all volumes used by the Pod.
+    // "OnVolumeMount" tries to mount volumes used by the Pod with the right context and skip recursive ownership
+    // change. Kubernetes may fall back to policy "Always" if a storage backed does not support this policy.
+    // This field is ignored for Pod's volumes that do not support SELinux.
+    // + optional
+    SELinuxRelabelPolicy *SELinuxRelabelPolicy
```

Comment:
What do you think about combining `SELinuxRelabelPolicy` and `FSGroupChangePolicy` into a single option? For example, something like this:

```go
const (
  // The heuristic policy acts like setting both the OnVolumeMount policy and the OnRootMismatch policy.
  HeuristicVolumeChangePolicy VolumeChangePolicy = "Heuristic"
  RecursiveVolumeChangePolicy VolumeChangePolicy = "Recursive"
)

type PodSeucrityContext struct {
  ...
  VolumeChangePolicy *VolumeChangePolicy
  ...
}
```

The motivation is that these settings seem very closely related, and would probably typically be set together. This decreases the flexibility, but simplifies the API and feature usage.

Reply:
We are targeting beta for `fsGroupChangePolicy` feature this quarter - https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/695-skip-permission-change . If we were to rename the field, we will have to probably put a halt on that. Nothing too problematic, I think it is fine if it stays alpha in one more release but just wanted to give heads up.

Reply:
Thinking through - there are some downsides of combining selinux and fsGroup options which are worth considering:

1. Having them behind one option kind of makes API hard to understand without reading the documentation in detail. 
2. There are use cases where one may want to set `OnRootMistach` but `Always` for selinux or vice-versa.  For example for a volume driver that supports selinux(such as gce-pd), using `OnVolumeMount` is perfectly fine default for all use cases even when you bring a volume with data on it. On other hand there could be cases where a user may have to use `Always` as default `fsGroupChangePolicy` but not `Always` selinux policy.

Reply:
I see benefits of a single field, however, "Heuristic" won't work well with `fsGroup` and a PV that's used also outside of Kubernetes, without `fsGroup` knowledge. If `Heuristic` was used, mounting `-o context` will work without issues, but skipping `fsGroup` `chown` / `chmod` may not work, because the top dir may have the right owner, while something outside of Kubernetes did create files on the volume with a wrong owner.

So, user may want full `fsGroup` change, but skip SELinux. I admit it may be a minor artificial use case, still, I don't want to paint us in the corner by combining the fields together.

Reply:
Added as "considered alternative".

Reply:
OK, this reasoning makes sense to me. It's unfortunate though, as the majority of the time I forsee the non-"always" option being used when the volume is too large and the recursive rewriting is too slow. It's unfortunate that these implementation details are getting surfaced in the API.

Reply:
In the SELinux case, would anyone want the Recursive approach over the VolumeMount approach? I'm wondering if can get rid of the SELinux policy in PodSecurityContext, and only use the CSIDriver field to figure out what to do.

Reply:
> In the SELinux case, would anyone want the Recursive approach over the VolumeMount approach

I can think only about shared volumes + subpaths, they behave a bit differently with `OnMount`.

BTW, there is still the first alternative - follow `fsGroup` / `OnRootMismatch`. Then we can merge the API easily, save **lot** of code and make everything consistent... Just the speed will suffer, as kubelet has to relabel the volume when used for the first time.

Reply:
The first alternative of having the same behavior of the fsgroup change policy sounds nice if that means we can also combine the API fields into one.

Reply:
I think the penalty of relabeling on the first mount is acceptable.